[BUG] Backup fails when Proactive Remediation has "/" character in Title

There appears to be an issue with backing up Proactive Remediations (and possibly other items) when a "/" character exists in the title as this is passed through to the file path for JSON/YAML backup (which is interpreted as a folder).

Backing up Proactive Remediation: Custom Device Inventory - App/Device Inventory
Traceback (most recent call last):
  File "/home/vsts/.local/bin/IntuneCD-startbackup", line 8, in <module>
    sys.exit(start())
  File "/home/vsts/.local/lib/python3.8/site-packages/IntuneCD/run_backup.py", line 241, in start
    run_backup(args.path, args.output, exclude, token)
  File "/home/vsts/.local/lib/python3.8/site-packages/IntuneCD/run_backup.py", line 193, in run_backup
    config_count += savebackup(path, output, exclude, token)
  File "/home/vsts/.local/lib/python3.8/site-packages/IntuneCD/backup_proactiveRemediation.py", line 72, in savebackup
    f = open(
FileNotFoundError: [Errno 2] No such file or directory: '/home/vsts/work/1/s/Proactive Remediations//Script Data/Custom Device Inventory - App/Device Inventory_DetectionScript.ps1'
##[error]Bash exited with code '1'.

--- a/README.md
+++ b/README.md
@@ -21,7 +21,8 @@ The package can also be run standalone outside a pipeline, or in one to only bac
 The front end for IntuneCD has now been released. Check it out [here](https://github.com/almenscorner/intunecd-monitor)
 
 ## What's new in 1.2.3
-- Bug fix where Proactive Remediations were not able to be backed up when containing a "/" in the display name.
+- New option to remove empty table rows from the documentation for a cleaner output. To enable this clean, just include `-c` when running `IntuneCD-startdocumentation`
+- Bug fix where Proactive Remediations were not able to be backed up when containing a "/" in the display name
 
 ## What's new in 1.2.2
 - Updated the Graph request module to handle `429` responses (throttling). If the response code is `429` the module will re-try the request after the amount of seconds specified in the `Retry-After` section in the response. If `429` is hit again, it will re-try until it succeeds. Example output,

--- a/README.md
+++ b/README.md
@@ -20,6 +20,9 @@ The package can also be run standalone outside a pipeline, or in one to only bac
 # Exciting news 📣
 The front end for IntuneCD has now been released. Check it out [here](https://github.com/almenscorner/intunecd-monitor)
 
+## What's new in 1.2.3
+- Bug fix where Proactive Remediations were not able to be backed up when containing a "/" in the display name.
+
 ## What's new in 1.2.2
 - Updated the Graph request module to handle `429` responses (throttling). If the response code is `429` the module will re-try the request after the amount of seconds specified in the `Retry-After` section in the response. If `429` is hit again, it will re-try until it succeeds. Example output,
   ```
@@ -41,11 +44,6 @@ The front end for IntuneCD has now been released. Check it out [here](https://gi
     - `IntuneCD-startbackup -e ConditionalAccess`
     - `IntuneCD-startupdate -e ConditionalAccess`
   - Assignments are currently not updated, changes to settings on policies are detected however and updated
-
-## What's new in 1.1.4
-- Bugfix where filters was not able to be updated with new values
-- Bugfix where notification templates was not able to be updated with new values
-- Update to how management intents are compared for diffs. Before there was a small chance it compared the wrong definitionId
 
 ## I use Powershell, Do I need to learn Python?
 No.

--- a/setup.cfg
+++ b/setup.cfg
@@ -1,6 +1,6 @@
 [metadata]
 name = IntuneCD
-version = 1.2.2
+version = 1.2.3
 author = Tobias Almén
 author_email = almenscorner@outlook.com
 description = Tool to backup and update configurations in Intune

--- a/src/IntuneCD/backup_proactiveRemediation.py
+++ b/src/IntuneCD/backup_proactiveRemediation.py
@@ -70,7 +70,7 @@ def savebackup(path, output, exclude, token):
                 decoded = base64.b64decode(
                     pr_details['detectionScriptContent']).decode('utf-8')
                 f = open(
-                    f"{configpath}/Script Data/{pr_details['displayName']}_DetectionScript.ps1",
+                    f"{configpath}/Script Data/{fname}_DetectionScript.ps1",
                     'w')
                 f.write(decoded)
                 # Save remediation script to the Script Data folder
@@ -78,7 +78,7 @@ def savebackup(path, output, exclude, token):
                 decoded = base64.b64decode(
                     pr_details['remediationScriptContent']).decode('utf-8')
                 f = open(
-                    f"{configpath}/Script Data/{pr_details['displayName']}_RemediationScript.ps1",
+                    f"{configpath}/Script Data/{fname}_RemediationScript.ps1",
                     'w')
                 f.write(decoded)
 

--- a/src/IntuneCD/documentation_functions.py
+++ b/src/IntuneCD/documentation_functions.py
@@ -151,7 +151,7 @@ def clean_list(data):
     return values
 
 
-def document_configs(configpath, outpath, header, max_length, split):
+def document_configs(configpath, outpath, header, max_length, split, cleanup):
     """
     This function documents the configuration.
 
@@ -204,6 +204,9 @@ def document_configs(configpath, outpath, header, max_length, split):
                 # Write configuration Markdown table
                 config_table_list = []
                 for key, value in zip(repo_data.keys(), clean_list(repo_data.values())):
+                    if cleanup:
+                        if not value:
+                            continue
                     if key == "@odata.type":
                         key = "Odata type"
                     else:

--- a/src/IntuneCD/run_documentation.py
+++ b/src/IntuneCD/run_documentation.py
@@ -52,10 +52,15 @@ def start():
         help='Split the documentation into multiple files and create index.md in the configpath directory with a '
              'list of all files',
     )
+    parser.add_argument(
+        '-c', '--cleanup',
+        help='If set, will remove all table rows with an empty value',
+        action='store_true'
+    )
 
     args = parser.parse_args()
 
-    def run_documentation(configpath, outpath, tenantname, jsondata,  maxlength, split):
+    def run_documentation(configpath, outpath, tenantname, jsondata,  maxlength, split, cleanup):
 
         now = datetime.now()
         current_date = now.strftime("%d/%m/%Y %H:%M:%S")
@@ -66,87 +71,87 @@ def start():
             md_file(outpath)
 
         # Document App Configuration
-        document_configs(f'{configpath}/App Configuration', outpath, 'App Configuration', maxlength, split)
+        document_configs(f'{configpath}/App Configuration', outpath, 'App Configuration', maxlength, split, cleanup)
 
         # Document App Protection
-        document_configs(f'{configpath}/App Protection', outpath, 'App Protection', maxlength, split)
+        document_configs(f'{configpath}/App Protection', outpath, 'App Protection', maxlength, split, cleanup)
 
         # Document Apple Push Notification
-        document_configs(f'{configpath}/Apple Push Notification', outpath, 'Apple Push Notification', maxlength, split)
+        document_configs(f'{configpath}/Apple Push Notification', outpath, 'Apple Push Notification', maxlength, split, cleanup)
 
         # Document Apple VPP Tokens
-        document_configs(f'{configpath}/Apple VPP Tokens', outpath, 'Apple VPP Tokens', maxlength, split)
+        document_configs(f'{configpath}/Apple VPP Tokens', outpath, 'Apple VPP Tokens', maxlength, split, cleanup)
 
         # Document iOS Applications
-        document_configs(f'{configpath}/Applications/iOS', outpath, 'iOS Applications', maxlength, split)
+        document_configs(f'{configpath}/Applications/iOS', outpath, 'iOS Applications', maxlength, split, cleanup)
 
         # Document macOS Applications
-        document_configs(f'{configpath}/Applications/macOS', outpath, 'macOS Applications', maxlength, split)
+        document_configs(f'{configpath}/Applications/macOS', outpath, 'macOS Applications', maxlength, split, cleanup)
 
         # Document Android Applications
-        document_configs(f'{configpath}/Applications/Android', outpath, 'Android Applications', maxlength, split)
+        document_configs(f'{configpath}/Applications/Android', outpath, 'Android Applications', maxlength, split, cleanup)
 
         # Document Windows Applications
-        document_configs(f'{configpath}/Applications/Windows', outpath, 'Windows Applications', maxlength, split)
+        document_configs(f'{configpath}/Applications/Windows', outpath, 'Windows Applications', maxlength, split, cleanup)
 
         # Document Web Apps
-        document_configs(f'{configpath}/Applications/Web App', outpath, 'Web Applications', maxlength, split)
+        document_configs(f'{configpath}/Applications/Web App', outpath, 'Web Applications', maxlength, split, cleanup)
 
         # Document Office Suite apps
         document_configs(f'{configpath}/Applications/Office Suite', outpath, 'Office Suite Applications', maxlength,
-                         split)
+                         split, cleanup)
 
         # Document compliance
-        document_configs(f'{configpath}/Compliance Policies/Policies', outpath, 'Compliance Policies', maxlength, split)
+        document_configs(f'{configpath}/Compliance Policies/Policies', outpath, 'Compliance Policies', maxlength, split, cleanup)
 
         # Message Templates
         document_configs(f'{configpath}/Compliance Policies/Message Templates', outpath, 'Message Templates', maxlength,
-                         split)
+                         split, cleanup)
 
         # Conditional Access
-        document_configs(f'{configpath}/Conditional Access', outpath, 'Conditional Access', maxlength, split)
+        document_configs(f'{configpath}/Conditional Access', outpath, 'Conditional Access', maxlength, split, cleanup)
 
         # Document profiles
-        document_configs(f'{configpath}/Device Configurations', outpath, 'Configuration Profiles', maxlength, split)
+        document_configs(f'{configpath}/Device Configurations', outpath, 'Configuration Profiles', maxlength, split, cleanup)
 
         # Document Group Policy Configurations
         document_configs(f'{configpath}/Group Policy Configurations', outpath, 'Group Policy Configurations', maxlength,
-                         split)
+                         split, cleanup)
 
         # Document Apple Enrollment Profiles
         document_configs(f'{configpath}/Enrollment Profiles/Apple', outpath, 'Apple Enrollment Profiles', maxlength,
-                         split)
+                         split, cleanup)
 
         # Document Windows Enrollment Profiles
         document_configs(f'{configpath}/Enrollment Profiles/Windows', outpath, 'Windows Enrollment Profiles', maxlength,
-                         split)
+                         split, cleanup)
 
         # Document Enrollment Status Page profiles
-        document_configs(f'{configpath}/Enrollment Profiles/Windows/ESP', outpath, 'Enrollment Status Page', maxlength, split)
+        document_configs(f'{configpath}/Enrollment Profiles/Windows/ESP', outpath, 'Enrollment Status Page', maxlength, split, cleanup)
 
         # Document filters
-        document_configs(f'{configpath}/Filters', outpath, 'Filters', maxlength, split)
+        document_configs(f'{configpath}/Filters', outpath, 'Filters', maxlength, split, cleanup)
 
         # Managed Google Play
-        document_configs(f'{configpath}/Managed Google Play', outpath, 'Managed Google Play', maxlength, split)
+        document_configs(f'{configpath}/Managed Google Play', outpath, 'Managed Google Play', maxlength, split, cleanup)
 
         # Document Intents
         document_management_intents(f'{configpath}/Management Intents/', outpath, 'Management Intents', split)
 
         # Document Partner Connections
-        document_configs(f'{configpath}/Partner Connections/', outpath, 'Partner Connections', maxlength, split)
+        document_configs(f'{configpath}/Partner Connections/', outpath, 'Partner Connections', maxlength, split, cleanup)
 
         # Document Proactive Remediations
-        document_configs(f'{configpath}/Proactive Remediations', outpath, 'Proactive Remediations', maxlength, split)
+        document_configs(f'{configpath}/Proactive Remediations', outpath, 'Proactive Remediations', maxlength, split, cleanup)
 
         # Document Shell Scripts
-        document_configs(f'{configpath}/Scripts/Shell', outpath, 'Shell Scripts', maxlength, split)
+        document_configs(f'{configpath}/Scripts/Shell', outpath, 'Shell Scripts', maxlength, split, cleanup)
 
         # Document Powershell Scripts
-        document_configs(f'{configpath}/Scripts/Powershell', outpath, 'Powershell Scripts', maxlength, split)
+        document_configs(f'{configpath}/Scripts/Powershell', outpath, 'Powershell Scripts', maxlength, split, cleanup)
 
         # Document Settings Catalog
-        document_configs(f'{configpath}/Settings Catalog', outpath, 'Settings Catalog', maxlength, split)
+        document_configs(f'{configpath}/Settings Catalog', outpath, 'Settings Catalog', maxlength, split, cleanup)
 
         if jsondata:
             json_dict = json.loads(jsondata)
@@ -196,7 +201,7 @@ def start():
                 l4 = f'{updated} {current_date} \n\n'
                 doc.writelines([l1, l2, l3, l4, document])
 
-    run_documentation(args.path, args.outpath, args.tenantname, args.jsondata, args.maxlength, args.split)
+    run_documentation(args.path, args.outpath, args.tenantname, args.jsondata, args.maxlength, args.split, args.cleanup)
 
 
 if __name__ == '__main__':

--- a/src/IntuneCD/update_conditionalAccess.py
+++ b/src/IntuneCD/update_conditionalAccess.py
@@ -57,7 +57,8 @@ def update(path, token):
                     print("-" * 90)
 
                     data['value']['conditions'].pop('users', None)
-                    data['value']['grantControls'].pop('authenticationStrength@odata.context', None)
+                    if data['value']['grantControls']:
+                        data['value']['grantControls'].pop('authenticationStrength@odata.context', None)
 
                     mem_id = data['value']['id']
                     # Remove keys before using DeepDiff

--- a/src/IntuneCD/update_proactiveRemediation.py
+++ b/src/IntuneCD/update_proactiveRemediation.py
@@ -16,6 +16,7 @@ from .check_file import check_file
 from .load_file import load_file
 from .remove_keys import remove_keys
 from .get_diff_output import get_diff_output
+from .clean_filename import clean_filename
 
 # Set MS Graph endpoint
 ENDPOINT = "https://graph.microsoft.com/beta/deviceManagement/deviceHealthScripts"
@@ -78,8 +79,9 @@ def update(path, token, assignment=False):
                     mem_data = remove_keys(mem_data)
 
                     # Check if script data is saved and read the file
-                    detection_script_name = f"{configpath}/Script Data/{repo_data['displayName']}_DetectionScript.ps1"
-                    remediation_script_name = f"{configpath}/Script Data/{repo_data['displayName']}_RemediationScript.ps1"
+                    fname = clean_filename(repo_data['displayName'])
+                    detection_script_name = f"{configpath}/Script Data/{fname}_DetectionScript.ps1"
+                    remediation_script_name = f"{configpath}/Script Data/{fname}_RemediationScript.ps1"
                     if os.path.exists(detection_script_name) and os.path.exists(
                             remediation_script_name):
                         with open(detection_script_name, 'r') as df:

--- a/tests/Update/test_update_conditionalAccess.py
+++ b/tests/Update/test_update_conditionalAccess.py
@@ -108,6 +108,18 @@ class TestUpdateConditionalAccess(unittest.TestCase):
         self.assertEqual(self.count, 0)
         self.assertEqual(self.makeapirequestPost.call_count, 0)
 
+    def test_update_config_grantControls(self):
+        """The count should be 0 and makeapirequestPost should not be called."""
+
+        self.repo_data['conditions']['applications']['excludedApplications'] = ['d4ebce55-015a-49b5-a083-c84d1797ae8c']
+        self.mem_data["value"][0]['grantControls'] = {'authenticationStrength@odata.context': 'test'}
+        self.makeapirequest.return_value = self.mem_data
+
+        self.count = update(self.directory.path, self.token)
+
+        self.assertEqual(self.count, 0)
+        self.assertEqual(self.makeapirequestPost.call_count, 0)
+
 
 if __name__ == '__main__':
     unittest.main()

--- a/tests/test_documentation_functions.py
+++ b/tests/test_documentation_functions.py
@@ -106,14 +106,16 @@ class TestDocumentationFunctions(unittest.TestCase):
             "config/test_file_name.json",
             '{"test": "test", "name": "test", "description": "test", "testvals": "1,2", "testlist": ["test"], "testdict": [{"test": "test"}], "assignments": [{"intent": "apply", "target": {"@odata.type": "#test", "groupName": "test-group", "deviceAndAppManagementAssignmentFilterId": "test-filter", "deviceAndAppManagementAssignmentFilterType": "test"}}]}',
             encoding="utf-8")
-        self.expected_data = '#test##testDescription:test###Assignments|intent|target|filtertype|filtername||------|----------|-----------|-----------||apply|test-group|test|test-filter||setting|value||--------|----------||Test|test||Name|test||Testvals|1<br/>2||Testlist|||Testdict|test:test|'
+        self.expected_data = '#test##testDescription:test###Assignments|intent|target|filtertype|filtername||------|----------|-----------|-----------||apply|test-group|test|test-filter||setting|value||--------|----------||Test|test||Name|test||Testvals|1<br/>2||Testdict|test:test|'
 
         document_configs(
             f"{self.directory.path}/config",
             f"{self.directory.path}/test.md",
             'test',
             10,
-            split=False)
+            split=False,
+            cleanup=True)
+            
         with open(f"{self.directory.path}/test.md", "r") as f:
             self.data = f.read()
             self.result = ''.join([line.strip() for line in self.data])

